### PR TITLE
Refractoring the ServiceLocator, closes #60

### LINKED ISSUE #60: Omit the use of Greenrobot.eventbus to distribute the ServiceLocator
- Parent issues: -
- Relevant issues: -
- Estimated effort: 3

* * *

## What
Remove the `ServiceLocator` + `ServiceLocatorCreater` construct where the classes that the `ServiceLocator` instantiates need to be told about the existence of the `ServiceLocator` via a communication bus.

## Why
Because it is incredibly cumbersome. Each of the classes mentioned above now requires three methods just for instantiation, and proper instantiation becomes very hard because the dependencies the `ServiceLocator` provides cannot be set to `final` attributes (since they can't be accessed in the constructor).

## How
My proposal is to remove the `ServiceLocatorCreator` entirely, make the `ServiceLocator` itself a singleton-like class and provide `this` when the `ServiceLocator` instantiates other classes.

## Definition of done
This is done once there is no longer any need for the greenrobot eventbus to distribute the ServiceLocator AND the ServiceLocator is provided via the constructor of relevant classes.

## Notes
n/a

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: #60 
- Related Pull Requests: ...

* * *

## What
This PR removes the ServiceLocatorCreator and gives modules the opportunity to access other modules over the service locator directly.

## Why
This is needed because we accessed unitialized modules in the constructor last week.

## How
If the app still runs, then it works fine ;-)

## Alternative implementation
I talked with Eric about many alternatives, and we agreed that this solutions would be the best.

## Notes
-

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 1]
```diff
@@ -40,47 +34,27 @@
 @SuppressWarnings("PMD.TooManyMethods")
 public final class GsonDatabaseAdapter implements IDatabase {
 
-    private final static Type TYPE_CONTACT_LIST = new TypeToken<ArrayList<Contact>>(){}.getType();
-    private final static Type TYPE_PROFILE_LIST = new TypeToken<ArrayList<Profile>>(){}.getType();
+    private final static Type TYPE_CONTACT_LIST = new TypeToken<ArrayList<Contact>>() {
+    }.getType();
+    private final static Type TYPE_PROFILE_LIST = new TypeToken<ArrayList<Profile>>() {
+    }.getType();
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Could you format these a little better 😇

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 2]
```diff
@@ -40,47 +34,27 @@
...
     private final static String CONTACT_NOT_FOUND = "Contact not found in database";
-    @SuppressWarnings("PMD.SingularField")
-    private final IServiceLocatorCreator serviceLocatorCreator;
-    private String contactsPath;
-    private String profilesPath;
+    @SuppressWarnings("PMD.SingularField") // This constants will be used later on probably
+    private final IConstants constants;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Just replace it with a local `final` variable in the constructor

File: app/src/main/java/com/nervousfish/nervousfish/service_locator/IServiceLocator.java [hunk 3]
```diff
@@ -25,4 +25,11 @@
     INFCHandler getNFCHandler();
     IQRHandler getQRHandler();
 
+    /**
+     * Registers the class specified to the EventBus
+     *
+     * @param object The Object that should be registed to the EventBus
+     */
+    void registerToEventBus(final Object object);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/service_locator/IServiceLocator.java]: Why is this needed, it is currently unused?

File: app/src/main/java/com/nervousfish/nervousfish/service_locator/ServiceLocator.java [hunk 4]
```diff
@@ -24,89 +35,108 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void registerToEventBus(final Object object) {
+        EventBus.getDefault().register(object);
+    }
+
+    private void checkModule(final IModule module, final String moduleName) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/service_locator/ServiceLocator.java]: Missing JavaDoc

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java (hunk 1 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: I'm sorry Eric, but no... I cannot I think :-8
If I change it back and do auto format in Intelllij, then it's changed back to this.

File: app/src/main/java/com/nervousfish/nervousfish/service_locator/IServiceLocator.java (hunk 3 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/service_locator/IServiceLocator.java]: Within a few days it will be used, for example to let the BluetoothActivity know that there were new Bluetooth devices detected or to let the NFC Activity know that some device touched your OnePlus 3

Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/service_locator/IServiceLocator.java]: 👍

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java (hunk 1 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: imo that shouldn't be a reason but fine 😕 Could you at least do it as follows to make it a little bit less ugly (if the execution line continues, it should start at a higher indentation)
```java
private final static Type TYPE_CONTACT_LIST =
            new TypeToken<ArrayList<Contact>>() {
            }.getType();
private final static Type TYPE_PROFILE_LIST =
            new TypeToken<ArrayList<Profile>>() {
            }.getType();
```
